File: daemons/linux/src/Codex/CodexReviewTarget.js
export function validReviewTarget(target) {
  if (!target || typeof target !== 'object' || Array.isArray(target)) { return false }
  const allowed = { uncommittedChanges: ['type'], baseBranch: ['type', 'branch'], commit: ['type', 'sha', 'title'], custom: ['type', 'instructions'] }[target.type]
  if (!allowed || Object.keys(target).some((key) => !allowed.includes(key))) { return false }
  if (target.type === 'uncommittedChanges') { return true }
  if (target.type === 'baseBranch') {
    return typeof target.branch === 'string' && target.branch.length > 0 && target.branch.length <= 255
      && !/[\s\u0000-\u001f\u007f~^:?*\[\]\\]/u.test(target.branch) && !target.branch.startsWith('-') && !target.branch.startsWith('/') && !target.branch.endsWith('/')
      && !target.branch.includes('..') && !target.branch.includes('//') && !target.branch.includes('@{')
      && target.branch.split('/').every((part) => !part.startsWith('.') && !part.endsWith('.') && !part.endsWith('.lock'))
  }
  if (target.type === 'commit') {
    return typeof target.sha === 'string' && /^[a-f0-9]{4,64}$/iu.test(target.sha)
      && (target.title === undefined || target.title === null || typeof target.title === 'string' && target.title.length <= 500 && !target.title.includes('\0'))
  }
  return typeof target.instructions === 'string' && target.instructions.trim().length > 0 && target.instructions.length <= 16000 && !target.instructions.includes('\0')
}

File: daemons/macos/src/CodexReviewTarget.swift
import Foundation

nonisolated enum CodexReviewTarget {
    static func valid(_ value: Any?) -> Bool {
        guard let target = value as? [String: Any], let type = target["type"] as? String,
            let allowed = [
                "uncommittedChanges": ["type"], "baseBranch": ["type", "branch"],
                "commit": ["type", "sha", "title"], "custom": ["type", "instructions"],
            ][type], Set(target.keys).isSubset(of: Set(allowed))
        else { return false }
        if type == "uncommittedChanges" { return true }
        if type == "baseBranch" {
            guard let branch = target["branch"] as? String else { return false }
            return !branch.isEmpty && branch.count <= 255
                && branch.range(of: #"[\s\x00-\x1f\x7f~^:?*\[\]\\]"#, options: .regularExpression) == nil
                && !branch.hasPrefix("-") && !branch.hasPrefix("/") && !branch.hasSuffix("/")
                && !branch.contains("..") && !branch.contains("//") && !branch.contains("@{")
                && branch.split(separator: "/", omittingEmptySubsequences: false).allSatisfy {
                    !$0.hasPrefix(".") && !$0.hasSuffix(".") && !$0.hasSuffix(".lock")
                }
        }
        if type == "commit" {
            guard let sha = target["sha"] as? String,
                sha.range(of: "^[a-fA-F0-9]{4,64}$", options: .regularExpression) != nil
            else { return false }
            if target["title"] == nil || target["title"] is NSNull { return true }
            return (target["title"] as? String).map { $0.count <= 500 && !$0.contains("\0") } == true
        }
        return (target["instructions"] as? String).map {
            !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && $0.count <= 16000 && !$0.contains("\0")
        } == true
    }
}
